### ISSUE TITLE
Event specification

### ISSUE THREAD

**Issue by l3pp4rd** [cucumber-attic/cucumber-pretty-formatter#2]
Hey @aslakhellesoy @mattwynne 

I've noticed that we have diverged in event specification. Lets make it clear and first agree on it completely for protocol version 0.1.0.

I've created a [readme.md](https://github.com/cucumber/cucumber-pretty-formatter/blob/refactor/README.md#events) with a list of **8** events and descriptions including parameters. I took some time to think and model it a little, added some additional information which seems to be necessary like matched step definition argument indices.

Please go through it and leave some comments or adapt it the way you think is more appropriate. My point is to nail down the event specification and I'm open for any discussion. If you have a different proposal, make a copy on wiki page as a proposal. But I hope comments may be enough.

I've been also thinking about other formats like excel, but currently at least for pretty printer I think that should be out of concern.

**Comment by jenisys** [comment]
I think it is good idea to have an event specification. This helps to clarify the common understanding when multiple people are working on it. Eventually, the event specification should be in an own file (IMHO), but for now it can be in "readme.md". 

A few suggestions:
- Each event attribute should also have a conceptual type (and I don't mean "string").
  For example, `TestCaseFinished.id` should have something like `FileLocation` as type.
- It is important to state what is required and what is optional.
  I would suggest that all attributes are required, if they are not marked as optional.
  Note that `TestRunFinished.memory` is an optional event attribute for me (but it is not described yet as optional).
- We need to define goals and non-goals for versions of the protocol. 
  For example, if the event protocol should support multiple test runs at the same time, an event attribute is missing in all events that also these events to be associated with this test run.

NOTE:
Eventually we need to write a documentation that will include the event specification. Because Sphinx currently seems to be the tool of choice for writing and publishing the documentation, it may be better to use a ReST `*.rst` document for the event specification.

**Comment by brasmusson** [comment]
To me `TestRunStarted`/`TestRunFinished` groups the whole execution, meaning within test cases from several feature files, and if suites are used, several suites are executed. When moving to compiling feature files to pickles/test case as done in Gherkin v3/v4 and Cucumber-Ruby v2, this is also a move away form the idea that "now this feature is executed". Also without using Cucumber-Ruby's `--random` feature, test cases from different features can be executed interleaved (see https://github.com/cucumber/cucumber-ruby/issues/952). Therefore:
- `TestRunFinished` cannot have an attribute `id` with the value `{path}:{line}`, or an attribute `suite`.

To me this means also that a formatter consuming the event stream need to be able to handle an event sequence like:
- `TestRunStarted`
- `GherkinSourceRead` (file1.feature)
- `GherkinSourceRead` (file2.feature)
- `<snip>`
- `TestCaseStarted` (from file1.feature)
- `<snip>`
- `TestCaseFinished`
- `TestCaseStarted` (from file2.feature)
- `<snip>`
- `TestCaseFinished`
- `<snip>`
- `TestRunFinished`

I think both `TestStepFinished` and `TestCaseFinished` should have a time attribute (both Cucumber-Ruby and Cucumber-JVM currently supply the duration of executed step definitions to formatters).

**Comment by l3pp4rd** [comment]
thanks @brasmusson indeed, the **TestRunFinished** event should not have an **id** and **suite** properties. Concerning test execution duration and memory used. Probably it would be appropriate to specify them in units instead of string format. For example a time could be in **milliseconds** and memory in **bytes**. maybe time could be renamed to **duration** and should be optional for all events with **Finished** suffix. Will update the specification based on that in a day.

**Comment by aslakhellesoy** [comment]
> Because Sphinx currently seems to be the tool of choice for writing and publishing the documentation, it may be better to use a ReST *.rst document for the event specification.

I agree. I suggest we eventually create a `specification/events` document under https://github.com/cucumber/docs for this. For now I think it's more convenient to let the event specification evolve in this repo - because this is currently the only implementation of the spec. Once it matures we can move it out.

**Comment by aslakhellesoy** [comment]
Regarding [GherkinSourceRead](https://github.com/cucumber/cucumber-pretty-formatter/blob/refactor/README.md#gherkinsourceread) I don't understand why it has a `suite` property. I don't think reading the source should be coupled to or related to execution. I think it belongs on `Test[Case|Step][Started|Finished]`.

I'd also like to propose renaming it to `TestSource` read - the type (Gherkin) can be inferred from the file extension - or we could have a `mimeType` property (with value `text/plain+gherkin` or something). This would allow for alternative formats such as Markdown without having to change or extend the spec.

Regarding time I think absolute values (timestamps) are better than durations. Durations can always be derived from two timestamps. The encoding of timestamps could be milliseconds since epoch.

Finally I think `location` is a more descriptive name than `id` - the value is indeed a location (line and/or column).

**Comment by l3pp4rd** [comment]
thanks for you input @jenisys @brasmusson @aslakhellesoy I've compiled your comments and updated the [event specification](https://github.com/cucumber/cucumber-pretty-formatter/blob/refactor/README.md#events).
- I somehow would prefer **duration** in milliseconds instead of start and end timestamps, I think the value for tests is duration it takes, not the time it starts. Maybe you have some arguments, what about other opinions?
- What about **suite** field in `StepDefinitionFound` event? Can suites have different step definitions for same step? Probably they can, @everzet does it make sense?

**Comment by aslakhellesoy** [comment]
Timestamps tell us _when_ scenarios ran. This opens up for some interesting reporting possibilities along the time axis.

**Comment by aslakhellesoy** [comment]
One thing that's missing is attachment events. (They're called embeddings in cucumber ruby/jvm but I prefer attachment).

This allows userland code (hooks and stepdefs) to attach extra information, such as a screenshot or even a video.

Attachments need the following properties:
- mime type (image/png)
- location (features/hello.feature:42)
- data (some encoded string)
- encoding (raw/base64/url)
- timestamp

In fact, stack traces could be represented as an attachment with mime type text/plain+stacktrace, encoding raw. This also removes the need for extra properties on StepFinished events.

Think of attachments as additional layers if decoration that is "attached" to a line in a file.

Some formatters will only handle some attachments and ignore the ones it cannot represent.

**Comment by brasmusson** [comment]
Test steps can either be based on steps from the test source (Gherkin), or they could come from (Before/After/AfterStep/Around) hooks. That means that the `location` in `TestStepStarted`/`TestStepFinished` either can refer to a line in a test source file or a line in a userland code file. 

Therefore I think that it would be better if the `StepDefinitionFound` had the attributes `step_location` and `definition_location` instead of `location` and `definition_id`. Both types of location can be used in the `TestStepStarted`/`TestStepFinished` later (even though a location in the userland code would be for a hook and not a step definition).

Maybe `TestStepStarted`/`TestStepFinished` should also have an attribute `step_type` (possible values: `step`, `before_hook` etc), or an optional attribute `hook_type` to be able to specify the type of hook.

**Comment by aslakhellesoy** [comment]
I like the sound of all of that @brasmusson

**Comment by l3pp4rd** [comment]
@aslakhellesoy attachments make sense, for html formatter that should be useful. Concerning **time** as Unix time like timestamp in milliseconds, fine.

@brasmusson I'm not familiar with `Around hook concept` is there a place somewhere I can read about it? OK, found it on wiki, nevermind

Regarding **definition_location** it might not necessarily be location, depends on implementation details. This field in general could be `optional` since it depends on implementation of cucumber, one might not provide hooks at all. It makes sense to have both of these fields, but **definition_id** or location as we agree to call it - I suggest to be optional. Also probably needs a better name? `source_ref` ?

`step_type` looks confusing to me, do we need to involve hooks into the formatter events?

**Comment by brasmusson** [comment]
By listening to the event stream it should be possible to know exactly what userland code was executed for a test case. Therefore the `StepDefinitionFound` event need to have an mandatory attribute to specify the userland code that was matched to the (gherkin) step. Therefore also hooks need to be reported through `TestStepStarted`/`TestStepFinished` events. This means that the `TestStepStarted`/`TestStepFinished` need to allow to specify either the location in a (gherkin) source file, or the userland code of a hook. If the `location` attribute `TestStepStarted`/`TestStepFinished` cannot be used to specify both, then these events need to have two optional attributes, `location` (for Gherkin steps) and `definition_location`/`definition_id`/`source_ref` (or what the attribute in `StepDefinitionFound` should be called - for the userland code of a hook).

**Comment by aslakhellesoy** [comment]
I'm at [cukeup](https://skillsmatter.com/conferences/7606-cukeup-2016) and just watched an interesting presentation from the BBC about their [HiveCI](http://bbc.github.io/hive-ci/).

What was particularly interesting is the format they have defined for test results - [Res](https://github.com/bbc/res).

We should definitely study it and see if there are useful concepts to borrow.

**Comment by mattwynne** [comment]
Terrific, thanks for writing up the spec @l3pp4rd!

TBH what you have there is not far away at all from what I've sketched out in Ruby Cucumber. I need a few more cycles to clean things up then I think we could start using it in this repo as a swap-in for `fake-cucumber` and test this thing out for real!

**Comment by l3pp4rd** [comment]
@aslakhellesoy I do not think **RES** is trying to achieve something similar, at least from the first look.
I probably will add missing parts as discussed this weekend and then start to implement printer.

**Comment by mattwynne** [comment]
Everyone also bear in mind that we can iterate on this protocol. Let’s try and get something working soon and then we can try to break it.

## 

The Cucumber logo is the intellectual property of Cucumber Ltd, a limited 
company registered in Scotland, number 456793.

UK Headquarters: Cucumber Ltd, Drumsyniebeg, Lochgoilhead, Cairndow, 
Argyll, PA24 8AN UK.

CONFIDENTIALITY NOTICE: The information in this e-mail is confidential and 
privileged; it is intended for use solely by the individual or entity named 
as the recipient hereof. Disclosure, copying, distribution, or use of the 
contents of this e-mail by persons other than the intended recipient is 
strictly prohibited and may violate applicable laws. If you have received 
this e-mail in error, please delete the original message and notify us by 
email immediately. Thank you. Cucumber Ltd.

**Comment by aslakhellesoy** [comment]
+1 for that. I've made some good progress on a cucumber-html-formatter
today (will push soon), and attacking the problem from several angles
simultaneously is the best way to make it good.

**Comment by mattwynne** [comment]
@l3pp4rd one other thing I think we should add to the event spec is a run_id (UUID) that we attach to all events. This will help when building central services that collate results from different boxes etc. WDYT?

**Comment by l3pp4rd** [comment]
OK, probably **run_id** should be optional. I'll update the event spec today based on previous comments.

**Comment by l3pp4rd** [comment]
Updated [event specification](https://github.com/cucumber/cucumber-pretty-formatter/blob/refactor/README.md#events)
- added **run_id** to every event
- added **TestAttachment** event
- changed durations, to timestamps on all `[started|finished]` events

@brasmusson it is still not clear to me about the **location-definition** reported by hooks. What does it mean, could you explain actions involved? If the step is executed directly by another step, maybe it should not report anything in that case since the step handles it's own state of execution, either it fails or not, no matter how many nested steps it executes by calling the step definitions directly.

**Comment by aslakhellesoy** [comment]
@l3pp4rd [hooks](https://github.com/cucumber/cucumber/wiki/Hooks) are not executed by other steps. They are weaved into scenarios, before and after gherkin steps (ignore around hooks, we're getting rid of those).

If you're familiar with xUnit tools, they're the same as before/after or setup/teardown methods.

They behave just like step definitions, but they don't have a gherkin counterpart. Therefore their location cannot point to a gherkin location.

**Comment by aslakhellesoy** [comment]
And - hooks have norhing to do with nested steps. That's another feature I'd like to get rid of. Totally pointless.

**Comment by l3pp4rd** [comment]
OK, got it @aslakhellesoy it relates only to before|after step hooks. guess scenario before|after hooks, does not impact on TestCase`Started|Finished` events

**Comment by brasmusson** [comment]
@l3pp4rd Running the `i18n/en` examples of the Cucumber-Ruby repo with the pretty formatter will include the output:

```
  Scenario: Regular numbers                  # features/division.feature:6
    * I have entered 3 into the calculator   # features/step_definitions/calculator_steps.rb:14
```

The text after `#` for the step is the `<path>:<line>` where the step definition matched for the step can be found. Using the event specification that text would come from the `definition_id` of the corresponding `StepDefinitionFound` event.

TestSteps based on Gherkin steps are both related to a location in a Gherkin feature file defining the Gherkin step they come from, and a location in a userland source code file where the matched step definition can be found.
TestSteps based on hooks are only related to a location in a userland source code file where the hook function/method can be found.

For a TestStep based on a Gherkin step it is possible to look up the location in a userland source code file where the matched step definition can be found (to be printed after the `#` in the output), based on the `location` and `suite` attribute of the `TestStepStarted`/`TestStepFinished` event and the `definition_id` of the `StepDefinitionFound` event the matching the `location` and `suite` attributes.
For TestSteps based on hooks, the only thing you have is the `TestStepStarted`/`TestStepFinished` events. Therefore they have to be able to define the location in a userland source code file where the hook function/method can be found. The (optional) attribute of the `TestStepStarted`/`TestStepFinished` events to do that would in a strong typing system have the same type as the `definition_id` of the `StepDefinitionFound`.

The pretty formatter (at least currently of Cucumber-Ruby and Cucmber-JVM) does not print any information about hooks. Unless they fail, then the stacktrace is printed. The event specification however needs still to supply the complete information about what was executed. The Json formatter (at least currently of Cucumber-Ruby and Cucmber-JVM) does print information also about hooks, so I use examples from the [spec of the Cucumber-Ruby Json formatter](https://github.com/cucumber/cucumber-ruby/blob/master/spec/cucumber/formatter/json_spec.rb#L672) below:

The data for a TestStep based on a Gherkin step (the path of the Gherkin file is defined for the feature as a whole and not for each step, so there is only a `line` attribute and not a `location` attribute):

```
  {"keyword": "Given ",
   "name": "there are bananas",
   "line": 4,
   "match": {"location": "spec/cucumber/formatter/json_spec.rb:669"},
   "result": {"status": "passed",
              "duration": 1}
  }
```

The data for a TestStep based on a hook. Since it is a hook there is not step test/name and no line number of the feature file, but is has `<path>:<line>` of the userland code executed same as the TestStep based on a Gherkin step has:

```
  {"match": {"location": "spec/cucumber/formatter/json_spec.rb:662"},
   "result": {"status": "passed",
              "duration": 1}
  }
```

**Comment by l3pp4rd** [comment]
> TestSteps based on Gherkin steps are both related to a location
> in a Gherkin feature file defining the Gherkin step they come from, and
> a location in a userland source code file where the matched step
> definition can be found.
> TestSteps based on hooks are only related to a location in a userland
> source code file where the hook function/method can be found.

My understanding of a test step case involves:
- before step hooks
- step execution
- after step hooks

The point is that all **xUnit** implementations treats it the same way,
whether the test fails or not, all must pass. Naturally, you can easily
wrap **step execution** function in a middleware chain and call it a hook.

If **test step case** fails, means either one of these 3 failed. It should
contain a readable error message so the user could understand what
happened. Was the error in hook or not and a stack trace if needed.

My opinion about **TestStep hooks** is the same, I do not think they
should be somehow related to event stream, but maybe mentioned in the
stack trace or error message only in case of failure.

@brasmusson the json formatter you linked, looks too complex. I do not
think that we should go the same road.

**Comment by aslakhellesoy** [comment]
I'm glad this is moving forward! I could probably answer some of the questions here, but in order to bring everyone involved to the same level of understanding I'd like to arrange a video conference instead. We could even do some [example mapping](https://cucumber.io/blog/2015/12/08/example-mapping-introduction) to explore this in more detail.

@mattwynne @tooky @jbpros and I are currently trying to spend most Fridays working on open source (UK business hours). What's people's availability in the next few Fridays?

**Comment by l3pp4rd** [comment]
I'm open this and next Friday from 5pm UK time

**Comment by brasmusson** [comment]
>  in order to bring everyone involved to the same level of understanding I'd like to arrange a video conference instead

Maybe at the coming "Show & Tell" (May 6, 2pm UK time)? It works for me.

**Comment by aslakhellesoy** [comment]
I'm travelling this Friday. What about my colleagues @tooky @mattwynne @jbpros - can any of you join?

**Comment by tooky** [comment]
I'm away too, sorry.

**Comment by mattwynne** [comment]
I'll be there. Can you make it @l3pp4rd?

**Comment by l3pp4rd** [comment]
2PM UK time, yes I most probably should, but that will be work hours and if something very important  happens I might be off.

**Comment by brasmusson** [comment]
In the Gherkin project specifies with the acceptance tests very clearly the structure and content of the AST and the compiled Pickles. However when it comes to Test Cases, which is what is actually executed by a Cucumber implementation, they are not covered by the Gherkin project. Cucumber-Ruby v2 does implement the concept of Test Cases that feature files are compiled to, but Cucumber-Ruby does not define a Json representation of Test Cases (which Gherkin does for the AST and the Pickles). Based on the [Pickles Json](https://github.com/cucumber/gherkin/blob/master/testdata/good/minimal.feature.pickles.json) of the [minimal.feature](https://github.com/cucumber/gherkin/blob/master/testdata/good/minimal.feature) in Gherkin, I have created a possible Json representation that feature compiled to a Cucumber-Ruby Test Case (assuming that there is one Before hook, one AfterStep hook and one After hook):

```
[
  {
    "locations": [
      {
        "column": 3,
        "line": 3,
        "path": "../testdata/good/minimal.feature"
      }
    ],
    "name": "minimalistic",
    "steps": [
      {
        "definition_location": {
          "line": 8,
          "path": "step_defintions/hooks.rb"
        },
        "text": "Before"
      },
      {
        "arguments": [],
        "definition_location": {
          "line": 6,
          "path": "step_defintions/steps.rb"
        },
        "locations": [
          {
            "column": 11,
            "line": 4,
            "path": "../testdata/good/minimal.feature"
          }
        ],
        "text": "the minimalism"
      },
      {
        "definition_location": {
          "line": 12,
          "path": "step_defintions/hooks.rb"
        },
        "text": "AfterStep"
      },
      {
        "definition_location": {
          "line": 16,
          "path": "step_defintions/hooks.rb"
        },
        "text": "After"
      }
    ],
    "tags": []
  }
]
```

Cucumber-Ruby does define one Test Step for each Gherkin/Pickles step and one Test Step for each loaded hook. The `locations` attribute (which refers to the line(s) in the feature file from which the Test Step is based) is optional - it is not applicable for Test Steps based on hooks. The `definition_location` attribute is also optional - it will be missing for undefined steps.

Based on this I do think that the event stream should include a TestStepStarted and a TestStepFinished also for each hook that is executed.

**Comment by l3pp4rd** [comment]
do we have a meeting today?

**Comment by tooky** [comment]
@l3pp4rd https://plus.google.com/hangouts/_/cucumber.pro/oss-core?authuser=0&hceid=Y3VjdW1iZXIucHJvX3NjcGV2NmtvcmxlcTgxZWRhbzZyMXRqdm00QGdyb3VwLmNhbGVuZGFyLmdvb2dsZS5jb20.jkl29f2uia7kdmcb497qnbcbb8

**Comment by l3pp4rd** [comment]
As discussed in the meeting, we will initially implement **progress** formatter to test the event specification. I will proceed with implementation on my free time. @mattwynne is the ruby event branch ready with the current spec? that would be helpful.

**Comment by mattwynne** [comment]
I'm sorry @l3pp4rd I haven't had any time to work on it yet.

I did come up with a great name for this repo though: **uni**versal **co**nsole w**r**iter AKA unicorn!

**Comment by l3pp4rd** [comment]
I like the name @mattwynne one thing though, is that it better should be unique, since most probably it will be installed in `$PATH`. I was thinking something with prefix **cu** if that make sense. There is [ruby http server - unicorn](https://unicorn.bogomips.org/) guess that may conflict. **cufmt** is the most straight forward name what comes to mind. I agree with unicorn if the ruby http server is not that common in practice.

**Comment by l3pp4rd** [comment]
I've made a basic event stream formatter in go cucumber implementation. Having more than one is much  better to test, since ruby implementation I think will differ in some aspects. Hopefully next week, I'll start the **unicorn** progress format implementation from event stream.

**Comment by mattwynne** [comment]
Good point about the binary name @l3pp4rd. I guess we could still call the project `cucumber-unicorn` and then figure out the right name for the binary later. `cunicorn` maybe? ;)

**Comment by l3pp4rd** [comment]
cunicorn is fine, maybe some other ideas will come up too. the most important is that it should not conflict. I think lets wait until it clears up and we come up with a best name, before starting to rename

**Comment by l3pp4rd** [comment]
probably we also need to attach snippets to **TestRunFinished** event:

``` json
{
    "event": "TestRunFinished",
    "status": "failed",
    "undefined_snippets": "plain text - optional",
    "timestamp": 1461436176456,
    "memory": 3456765,
    "run_id": "uuid"
}
```
